File: riddles/problem solving/medium.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <math.h>
using namespace std;

string solve(string &s)
{
    vector<int> num;
    vector<string> str;
    string temp = "";
    for (int i = 0; i < (int)s.size(); i++)
    {
        if (s[i] >= '0' && s[i] <= '9')
        {
            int n = 0;
            while (s[i] >= '0' && s[i] <= '9')
            {
                n = n * 10 + (s[i] - '0');
                i++;
            }
            i--;
            num.push_back(n);
        }
        else if (s[i] == '[')
        {
            str.push_back(temp);
            temp = "";
        }
        else if (s[i] == ']')
        {
            string t = str.back();
            str.pop_back();
            int n = num.back();
            num.pop_back();
            for (int j = 0; j < n; j++)
            {
                t += temp;
            }
            temp = t;
        }
        else
        {
            temp += s[i];
        }
    }
    return temp;
}
int main()
{
    string s = "3[a4[b]3[c]]2[a]"; // out delldelldell
    string out = solve(s);
    cout << out << endl;
    return 0;
}

File: riddles/problem solving/medium.py
def solve(s):
    num = []
    str_list = []
    temp = ""
    
    i = 0
    while i < len(s):
        if '0' <= s[i] <= '9':
            n = 0
            while '0' <= s[i] <= '9':
                n = n * 10 + int(s[i])
                i += 1
            i -= 1
            num.append(n)
        elif s[i] == '[':
            str_list.append(temp)
            temp = ""
        elif s[i] == ']':
            t = str_list.pop()
            n = num.pop()
            temp = t + temp * n
        else:
            temp += s[i]
        i += 1
    
    return temp

if __name__ == "__main__":
    s = "3[a4[b]3[c]]2[a]"  # out delldelldell
    out = solve(s)
    print(out)
